### DISCUSSION TITLE
RFC: Fix block rewards

### DISCUSSION THREAD

**Discussion by aesedepece** [witnet/witnet-rust#2239] [Ideas]
# Where are we?
- Minting rewards and [coin issuance](https://github.com/witnet/WIPs/blob/master/wip-0003.md) in Witnet work much alike to how they do in Bitcoin: 
  - A block proposer is allowed to take a fixed amount of freshly minted coins (250 Wits), plus all unassigned value from the transactions contained in the block, aka _fees_.
  - There exist a halving: that fixed amount of coins that are freshly minted in every block is reduced periodically. In the case of Witnet, that's set to happen every 3500000 epochs, i.e. a first halving will take place in October 2025.
- The whole point of having a programatic issuance mechanism is to incentivize the bootstrapping of the network — make it attractive for people to operate nodes from day one even if there's little demand for block space.
-  That "subsidization" of node operation through issuance is expected to buy the whole ecosystem enough time to perform price discovery of its oracle services. That is, miners don't need to charge much to the end users for processing their oracle queries, because they are already earning coins through block minting.
- In this stage, without that strong of a need to incentivize the miners to process the oracle queries, the fees paid for those merely buy "priority". Miners are expected to prioritize oracle queries in their mempool based on how much fees they pay vs. how much block space they take, so as to maximize profit.
- However, halvings are there to retire the subsidization slowly over time, and progressively let the fees make for an ever increasing percentage of the miner's revenue.
- That would in theory tend to make oracle queries more expensive. But in practice, if right after one of the halvings the mining activity becomes loss-making for a node operator, they have no way to "ask for more fees". They will simply stop proposing blocks, and some other miner's blocks will be accepted, increasing that other miner's profit. This model based in economic equilibrium has been more than proven by Bitcoin since 2009.
- Moreover, the halving period is so long, that the long term tendency of the mining activity in Witnet is to reach commoditization — the operating cost for running a mining node can only go down over time (you can't say the same of PoW mechanisms, because the cost of the energy can spike for many reasons).

# What's the problem?
- Full disclosure: there's no real problem!
- But let's reflect on this: if miners are perfectly OK with minting a block for 250 Wits, and they are not expecting transaction fees (as they mint empty blocks once in a while), and transaction fees are there only to buy priority, then there's no point in having a fixed issuance and putting the transaction fees on top.
- We can say that putting the fees on top only leads to unjustified inflation that misses the original point of the issuance model, which is to make up for the low transaction volume in the early days of the protocol.

# Can devs do something?
- Other chains have successfully introduced changes in their issuance models (e.g. [EIP-1559](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md) in Ethereum) with the expectation of counterbalancing inflation without affecting the miner's expectation for rewards. Those changes have had a positive impact on both on the fees market and in [limiting the inflation](https://watchtheburn.com/), with no compromise or backlash from the miners.

# My proposal
- We can simply fix the block reward, and then only issue that amount minus the fees. E.g. if a block collects 10 Wits in fees, the issuance will be 240 Wits, and the total reward for the miner will still be 250 Wits.
- May the fees be equal or greater than the targeted block reward, the miner will take them all, but no new coins will be minted.
- Halvings would still work the same, but applied on block rewards instead of on issuance.
- Over time, this would effectively reduce the circulating supply in comparison to the expected issuance curve in the same amount as the Wit used for transaction fees. That is, the more Witnet is used, the lesser the inflation!

**Comment by guidiaz** [comment]
It really looks like a great idea. Although I have one concern with respect to the fact that the ultimate circulating supply would not be deterministic, as the inflation rate on every halving period would depend on the average amount of data request fees per block during that period. 

I'm not sure whether this could be an issue or not, but if it were, we could then consider to halve the "maximum" block reward not depending on a pre-established period of time, but on a pre-established inflation rate per halving "era". This way, the ultimate circulating supply would depend on pre-established inflation rates per era, and the expected time for the circulating supply to reach its maximum would ultimately depend on how much Witnet keeps being used for trustlessly solving data requests.

**Reply by aesedepece** [reply]
> [...] I have one concern with respect to the fact that the ultimate circulating supply would not be deterministic [...]

At this point I see this as a no-problem because:
- No protocol stakeholder is negatively impacted by such a change (quite the opposite indeed!)
- Miners in particular should be fine with it because the foundation for this proposal is the fact that they are perfectly OK with minting a block for 250 Wits, regardless of those coming from issuance or from fees. In fact, again, if the coins come from fees it's even better for them, because they don't get diluted through inflation, and the more their income relies on fees, the better, because the coins are being used in the network instead of flooding the market.
- It doesn't harm the simplicity and elegance of the protocol either, as the maximum supply has been always nondeterministic because of rollbacks and burning.

> we could then consider to halve the "maximum" block reward not depending on a pre-established period of time, but on a pre-established inflation rate per halving "era"

Aha, I see. I think you are suggesting that every time we enter a new era, the block reward is "retargeted" taking into account the actual issuance from the previous eras, so that the issuance curve remains unchanged in the big picture. That sounds very reasonable to me, but I tend to think it's not worth the effort because of my points above.

**Comment by drcpu-github** [comment]
I am not in favor of this proposal for a multitude of reasons. First let me break down the initial post and highlight where I disagree.

> That "subsidization" of node operation through issuance is expected to buy the whole ecosystem enough time to perform price discovery of its oracle services. That is, miners don't need to charge much to the end users for processing their oracle queries, because they are already earning coins through block minting.

I'd argue we have not seen proper price discovery of the oracle services yet. Adoption of the Witnet oracle is only just starting to take off and most blocks are not actually full. Hence we don't know how much users are willing to pay for data requests (as they can simply set their fees to the lowest possible value) nor what the lowest price is miners are willing to accept when constructing blocks (since they currently should accept all transactions in order to maximize their earnings). Furthermore, almost all data requests are currently being routed over the centralized bridge operated by the Witnet Foundation whom impose an artificial limit on the rewards of a data request. Until there are more bridge operators and the ecosystem has evolved more, I'd argue that we cannot state proper price discovery has happened.

> But let's reflect on this: if miners are perfectly OK with minting a block for 250 Wits, and they are not expecting transaction fees (as they mint empty blocks once in a while), and transaction fees are there only to buy priority, then there's no point in having a fixed issuance and putting the transaction fees on top.

Over the past two years, many miners have come and left again. Most of the miners I talked to whom quit said they stopped mining because they could not achieve a positive ROI. Obviously this ties in to your argument an economic equilibrium, but I think it foregoes the fact that many miners currently seem to be either mine out of ideological reasons (e.g., to secure and help grow the network) and / or because they believe the future price will compensate their current costs. In both cases I find it strange to state that means they would be fine with a reduced block reward.

I checked the number of blocks that did not contain any transactions. Since epoch 1.000.000, about 7 months ago, only 10.718 out of  430.664 blocks were empty (close to 2.5%). I don't think drawing a conclusion about the average miner not expecting transaction fees based on such a small percentage of total produced blocks makes a lot of sense.

Moreover, miners can mine empty blocks for a multitude of reasons:

1. Their node(s) is (are) configured badly
2. They don't want to include certain types of transactions
3. They have bad luck and are connected to bad peers resulting in their transaction pool being empty from time to time
4. Some other miner is attempting to censor them.

It seems strange to group all those reasons under the denominator of miners don't expect transaction fees so let's change the issuance schedule.

> Other chains have successfully introduced changes in their issuance models (e.g. EIP-1559 in Ethereum) with the expectation of counterbalancing inflation without affecting the miner's expectation for rewards. Those changes have had a positive impact on both on the fees market and in limiting the inflation, with no compromise or backlash from the miners.

If I recall correctly, there actually was significant opposition of (a subset of) miners. However, the users of the Ethereum network viewed this update as an improvement over the state of the network at that time, and thus based on the social consensus miners eventually followed suit as they considered a reduction in earned fees better than no fees at all.

This does bring me to the main issue I have with this proposal: I  think it misses a sound (game-)theoretical justification. It seems like burning for the sake of reducing inflation without a sound underlying reason. To illustrate what I mean: as far as I understood the reason for implementing EIP-1559 was to make the transaction fees more predictable and burning the base fee served the goal of making it economically uninteresting for miners to manipulate the overall transaction fees. The fact that it reduces inflation is a side-effect. I feel like a justification like with EIP-1559 is missing in this proposal.

Furthermore, at the moment, the fees of transactions included in a block are negligible. This proposal would not even make a dent in the overall inflation. I don't think making a far reaching decision like this based on the current state of block rewards versus fees is the way forward. This can obviously change with adoption, and at that time this proposal could be reevaluated.
